overlord/snapstate/autorefresh_gating_test.go

```diff
@@ -211,6 +223,392 @@ slots:
+	// snap-b was refreshed a few days ago
+	mockLastRefreshed(c, st, "2021-05-01T10:00:00Z", "snap-b")
+
+	// holding itself
+	hold := time.Hour * 24 * 96
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", hold, "snap-a"), ErrorMatches, `cannot hold some snaps:\n - requested holding duration 2304h0m0s exceeds maximum holding for snap "snap-a"`)
+
+	// holding other snap
+	hold = time.Hour * 49
+	err := snapstate.HoldRefresh(st, "snap-a", hold, "snap-b")
+	c.Check(err, ErrorMatches, `cannot hold some snaps:\n - requested holding duration 49h0m0s exceeds maximum holding for snap "snap-b"`)
+	herr, ok := err.(*snapstate.HoldError)
+	c.Assert(ok, Equals, true)
+	c.Check(herr.SnapsInError, DeepEquals, map[string]snapstate.HoldDurationError{
+		"snap-b": {
+			Err:          fmt.Errorf(`requested holding duration 49h0m0s exceeds maximum holding for snap "snap-b"`),
+			DurationLeft: 48 * time.Hour,
+		},
+	})
+
+	// hold for maximum allowed for other snaps
+	hold = time.Hour * 48
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", hold, "snap-b"), IsNil)
+	// 2 days passed since it was first held
+	now = "2021-05-12T10:00:00Z"
+	hold = time.Minute * 2
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", hold, "snap-b"), ErrorMatches, `cannot hold some snaps:\n - snap "snap-a" cannot hold snap "snap-b" anymore, maximum refresh postponement exceeded`)
+
+	// refreshed long time ago (> maxPostponement)
+	mockLastRefreshed(c, st, "2021-01-01T10:00:00Z", "snap-b")
+	hold = time.Hour * 2
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-b"), ErrorMatches, `cannot hold some snaps:\n - snap "snap-b" cannot hold snap "snap-b" anymore, maximum refresh postponement exceeded`)
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", 0, "snap-b"), ErrorMatches, `cannot hold some snaps:\n - snap "snap-b" cannot hold snap "snap-b" anymore, maximum refresh postponement exceeded`)
+}
+
+func (s *autorefreshGatingSuite) TestHoldAndProceedWithRefreshHelper(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+	mockInstalledSnap(c, st, snapCyaml, false)
+	mockInstalledSnap(c, st, snapDyaml, false)
+
+	mockLastRefreshed(c, st, "2021-05-09T10:00:00Z", "snap-b", "snap-c", "snap-d")
+
+	restore := snapstate.MockTimeNow(func() time.Time {
+		t, err := time.Parse(time.RFC3339, "2021-05-10T10:00:00Z")
+		c.Assert(err, IsNil)
+		return t
+	})
+	defer restore()
+
+	// nothing is held initially
+	held, err := snapstate.HeldSnaps(st)
+	c.Assert(err, IsNil)
+	c.Check(held, IsNil)
+
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", 0, "snap-b", "snap-c"), IsNil)
+	c.Assert(snapstate.HoldRefresh(st, "snap-d", 0, "snap-c"), IsNil)
+	// holding self
+	c.Assert(snapstate.HoldRefresh(st, "snap-d", time.Hour*24*4, "snap-d"), IsNil)
+
+	held, err = snapstate.HeldSnaps(st)
+	c.Assert(err, IsNil)
+	c.Check(held, DeepEquals, map[string]bool{"snap-b": true, "snap-c": true, "snap-d": true})
```

Comment:
this test is confusing to me, I must have misread something, I interpreted the above HoldRefresh calls as setting up state thusly:

- snap-a is being held back by snap-b and snap-c
- snap-d is being held back by snap-c
- snap-d is being held back by itself

Which then, a function returning "held" snaps that should not be refreshed, I would expect to return `snap-a` and `snap-d`. 

Why is `snap-a` not in this map, and why is `snap-b` and `snap-c` in this map? It seems like this result is actually the set of snaps _that are holding back other snaps_, i.e. snaps that are gating/preventing other snaps from being refreshed, but the doc-comment says that this function returns snaps that are being held back.

I must have missed something somewhere...

Reply:
It's the other way around :). snap-a is the gating snap and it holds snap-b and snap-c  and so on. Please see the other comment where I explained the semantics some more.